pkg/util/clienthelper/clienthelper.go

```diff
@@ -34,35 +33,34 @@ import (
 	_ "github.com/Azure/ARO-RP/pkg/util/scheme"
 )
 
-type Interface interface {
-	EnsureDeleted(ctx context.Context, gvk schema.GroupVersionKind, key types.NamespacedName) error
+type Writer interface {
+	client.Writer
+	// Ensure applies self-contained objects to a Kubernetes API, merging
```

Comment:
I know you're not changing this in the PR, so this is not a blocker here, but we should use server-side apply instead of whatever client-side logic exists, as it's guaranteed to be a better implementation :)

Reply:
I would _love_ to use server-side apply, but the fake implementation of controller-runtime doesn't yet support it :(

Refs: 
- https://github.com/kubernetes-sigs/controller-runtime/issues/2341
- https://github.com/kubernetes/kubernetes/pull/125560 -- the groundwork does look like it's beginning to be there :)
- https://github.com/kubernetes-sigs/controller-runtime/pull/2702 hi @troy0820 fancy seeing you there :D

Reply:
_No comment, just waving at_ @hawkowl 👋🏽

Reply:
Yes, the controller-runtime fakes are horrendous and IMO moving away from them as soon as possible is also a great boon to the project :)

SSA is very complex and faking it is likely not going to happen - not even just in some soon time-frame, but perhaps ever. Client fakes have fallen out of favor and everyone has a horror story of working around the mess that is envtest without gaining any confidence in their application functioning in the real world against a real API server. Even past that, the concerns Alvaro has in Troy's PR are a mountain of work that the upstream SIGs are not interested in taking up or sponsoring, from my understanding.

This is neither here nor there, but you might want to consider not putting SSA support in the controller-runtime fakes as a precondition to using this (amazingly useful and beneficial) technology in your production deployments.
 
See an example of how to test controllers or similar using static data in, static data out here: https://github.com/openshift/hypershift/blob/main/control-plane-pki-operator/certificaterevocationcontroller/certificaterevocationcontroller_test.go

Reply:
The issue beyond the requisite work are the apply types that are not scaffold with crd/operator-sdk/kubebuilder.  That work is a side effect of the client (controller-runtime) not implementing that whereas client-go has that necessary logic.  

The interface client being used here is the mechanism where that can't be utilized. I have no context around what this is trying to accomplish but if it is indeed using corev1 types where a `kubernetes.Interface` can be used, those applyConfigurations are available to fulfil the server side apply.

The faking of the clients are tracked with a "tracker" and trying to track a server side apply in any client would be trying to hold state for things it wouldn't worry about because of the client's responsibility.  Having a test suite where that can happen seems outside of the scope of what the client/fakes could do

Reply:
I have worked with the `controller-runtime` ecosystem and upstream Kubernetes client-sets for many years, so I'm pretty familiar with the structure. I don't think I communicated the "static data in, static data out" link well - my suggestion is not that some client helper hold on to apply configurations, nor is it to somehow teach the `controller-runtime` client how to apply. 

In case we get too far in the weeds - I just want to re-iterate that the comments here are not review comments for this PR, just a general comment. I'm not suggesting any changes are made right now.

---

What I *am* suggesting, though, is that operator reconciliation loops are written as pure functions - either take some set of cluster state as input, or allow dependency injection to take producers of cluster state as input, and produce state as output. Apply configurations have a wonderful property in that they encode your intent in static data - so, testing your lower-case `r` `reconcile()` is straightforward, and no mocks of any persuasion are necessary.

The top-level `Reconcile()` might look like:

```go
func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
    // assemble the data necessary by using listers, etc
    input := r.getInput()
    
    // transform input state into intended state
    intent, err := r.reconcile(input)
    
    // realize intended state onto the cluster
    return r.execute(intent)
}
```

Where `r.execute(intent)` takes apply configurations and uses a `kubernetes.Clientset` to call `Apply`.

--- 

In my experience, relying heavily on mocks leads many teams to stumble into deeply frustrating issues, often many years down the line. Examples I've seen:

- implementing highly-performant caching using metadata-only watches requires 100% of tests to be rewritten, as the fakes don't know how to propagate data between the metadata fake and the "real" fake
- simplifying controller code to reduce maintenance burden and eliminate incorrect client-side patching by using server-side apply means all unit tests fail
- fragile assumptions about the number of times your reconciliation is called and when, in what order, etc, lead integration tests to be a poor substitute for end-to-end testing against actual API servers, leading to test duplication
- fakes make many highly surprising decisions, taking valuable mental overhead and productivty away from the team, like:
  - faking out field selectors (usually handled in the server) with indices, which requires test code to be written to create indices
  - objects with deletion timestamps but no finalizers cause some internal panic even though it's totally valid for your controller to see one in production
  - fake clients need subresources registered manually
  
---

Teams that spend time writing unit tests for unit-level concerns and do not reach to mocks, in my experience, move faster and get a higher RoI on their tests. End-to-end tests and chaos tests validate production modalities that integration tests attempt to, but cannot.

Reply:
@stevekuznetsov I completely agree with that. 🥇 

Even though I am by far the one less experienced in this conversation about this domain, I have seen similar problems in other domains that made me resonate a lot with your comment while reading it.

Sometimes we focus too much on trying to reproduce the state of the world where our app will run, and we forget that we should primarily test what our app does, and what our business logic is. It is almost always much simpler and has a greater return on investment to just have our business logic in pure functions that just expect input data and produce output data when talking about unit tests. We then forget about how this data will be gathered and what will be done with the result, and we can simply ensure that domainLogicFn(x) == expectedOutput.

This is a common concern that applies to a lot of different domains, but it is pretty common in k8s because most of the time the business logic gets mixed/loosed in the effort to understand/simulate the "state" where all that business logic will happen (due to the stateful nature of k8s controllers). So I just want to reaffirm that avoiding complex mocks when possible with pure functions is a great idea and not just in this context 🙂